treeherder/model/data_cycling/cyclers.py

```diff
@@ -130,15 +135,33 @@ def cycle(self):
             logger.warning(ex)
 
     def _remove_leftovers(self):
-        # remove any signatures which are
-        # no longer associated with a job
+        self.__remove_empty_signatures()
+
+        self.__remove_too_old_alerts()
+        self.__remove_empty_alert_summaries()
+
+        self.__remove_empty_backfill_reports()
+
+    def __remove_empty_signatures(self):
+        logger.warning("Removing performance signatures which don't have any data points...")
         signatures = PerformanceSignature.objects.filter(last_updated__lte=self.max_timestamp)
```

Comment:
This filters empty signatures? We are sure they are empty because time passed and data points have been surely deleted?

Reply:
Not exactly. We do a very basic filtering here. We're fetching all signatures that are candidates for removal.
The `PublicSignatureRemover` will do further filtering & delete only those that don't have any more data or are from the try repository.